src/lib/support/Pool.h

```diff
@@ -331,6 +394,86 @@ class HeapObjectPool : public internal::Statistics, public HeapObjectPoolExitHan
 #endif // __SANITIZE_ADDRESS__
     }
 
+    /// Provides iteration over active objects in the pool.
+    ///
+    /// NOTE: There is extra logic to allow objects release WHILE the iterator is
+    ///       active while still allowing to advance the iterator.
+    ///       This is done by flagging an iteration depth whenever an active
+    ///       iterator exists. This also means that while a pool iterator exists, releasing
+    ///       of tracking memory objects may be deferred until the last active iterator is
+    ///       released.
+    class ActiveObjectIterator
+    {
+    public:
+        using value_type = T;
+        using pointer    = T *;
+        using reference  = T &;
+
+        ActiveObjectIterator() {}
+        ActiveObjectIterator(const ActiveObjectIterator & other) : mCurrent(other.mCurrent), mEnd(other.mEnd)
+        {
+            if (mEnd != nullptr)
+            {
+                mEnd->mIterationDepth++;
+            }
+        }
+
+        ActiveObjectIterator & operator=(const ActiveObjectIterator & other)
+        {
+            if (mEnd != nullptr)
+            {
+                mEnd->mIterationDepth--;
+                mEnd->CleanupDeferredReleases();
+            }
+            mCurrent = other.mCurrent;
+            mEnd     = other.mEnd;
+            mEnd->mIterationDepth++;
+        }
+
+        ~ActiveObjectIterator()
+        {
+            if (mEnd != nullptr)
+            {
+                mEnd->mIterationDepth--;
+                mEnd->CleanupDeferredReleases();
+            }
+        }
+
+        bool operator==(const ActiveObjectIterator & other) const
+        {
+            // extra current/end compare is to have all "end iterators"
+            // compare as equal (in particular default active object iterator is the end
+            // of an iterator)
+            return (mCurrent == other.mCurrent) || ((mCurrent == mEnd) && (other.mCurrent == other.mEnd));
+        }
+        bool operator!=(const ActiveObjectIterator & other) const { return !(*this == other); }
+        ActiveObjectIterator & operator++()
+        {
+            do
+            {
+                mCurrent = mCurrent->mNext;
+            } while ((mCurrent != mEnd) && (mCurrent->mObject == nullptr));
+            return *this;
+        }
+        T * operator*() const { return static_cast<T *>(mCurrent->mObject); }
+
+    protected:
+        friend class HeapObjectPool<T>;
+
+        explicit ActiveObjectIterator(internal::HeapObjectListNode * current, internal::HeapObjectList * end) :
```

Comment:
So why is this using a protected ctor and friend decl while the non-heap version does not.  Seems like the non-heap version should too.

Reply:
I made minimum changes to make things compile. The class are not well split. I will take a look, however right now I believe the semantics between bitmap and heap are different, so their iterators are different.

Reply:
The trouble for BitmapObjectPool was that it takes another size argument `N` so I cannot fried like HeapObjectPool<T> (where I cover everything) but instead I would need friend BitmapObjectPool<T, N> for all N, which I could not get to be accepted cleanly so I gave up.

We can have some followups if needed, however I am not sure if the extra work will be worth it.